core/src/poh_recorder.rs

```diff
@@ -136,6 +136,33 @@ impl PohRecorder {
         self.ticks_per_slot
     }
 
+    fn received_any_previous_leader_data(&self, slot: Slot) -> bool {
+        (slot.saturating_sub(NUM_CONSECUTIVE_LEADER_SLOTS)..slot).any(|i| {
+            // Check if we have received any data in previous leader's slots
+            if let Ok(slot_meta) = self.blocktree.meta(i as Slot) {
+                if let Some(slot_meta) = slot_meta {
+                    slot_meta.received > 0
+                } else {
+                    false
+                }
+            } else {
+                false
+            }
+        })
+    }
+
+    fn reached_leader_tick(&self, leader_first_tick_height: u64) -> bool {
+        let target_tick_height = leader_first_tick_height.saturating_sub(1);
+        let ideal_target_tick_height = target_tick_height.saturating_sub(self.grace_ticks);
+        let current_slot = self.tick_height / self.ticks_per_slot;
+        // we've approached target_tick_height OR poh was reset to run immediately
+        // Or, previous leader didn't transmit in any of its leader slots, so ignore grace ticks
+        self.tick_height >= target_tick_height
+            || self.start_tick_height + self.grace_ticks == leader_first_tick_height
```

Comment:
Quick clarification (I know this was in the original code as well)

If you have `N` consecutive slots, won't `self.start_tick_height + self.grace_ticks == leader_first_tick_height` return true for every slot after the first one? (I think self.start_tick_height is set on `reset` which is the beginning of the first slot). 

Will this function then return true every single time ReplayStage checks `reached_leader_slot` after the first slot? Is this ok b/c this check in ReplayStage will make sure another bank for the same slot isn't created: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L482

Reply:
It should be ok due to this check https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L377

If the leader already has bank, it's not calling `maybe_start_leader` which in turn calls this function.
WDYT?

Reply:
@pgarg66 I think it's possible for tick() -> flush() here: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L377 to clear the bank, before ReplayStage has reset to a different bank, and then that check won't prevent a call to `maybe_start_leader`. But then this check should catch that case: https://github.com/solana-labs/solana/blob/master/core/src/replay_stage.rs#L482.

Reply:
cool. so it's all good? anyway, I'll merge this PR. If this is still an issue, we can track it as a separate issue.

Reply:
yup, looks like it, thanks!